Fix some api tests (WIP)

crookedneighbor opened the pull request:
Could someone check my work?

Many of the task counts were off because of adding additional sample tasks.

Remove the registerNewUser function from before the Groups test, because it wasn't testing against the right user.

Added a todo for the quest tests, invites weren't being saved.

Someone set subscription tests to skip, but the it block was indented incorrectly

crookedneighbor committed:
Corrected tests for archiving todos to reflect the correct number of default todos when creating account.

crookedneighbor committed:
Changed challenges completed count to calculate from total number of user tasks minus 2 (the two completed)

crookedneighbor committed:
Removed the create a new user function call from groups, as it was now using the correct user to look up the leader

crookedneighbor committed:
Wrote a todo for test

crookedneighbor commented on test/api.mocha.coffee:
```diff
@@ -161,16 +161,15 @@ describe "API", ->
     describe "Groups", ->
       group = undefined
       before (done) ->
-        registerNewUser ->
```
Anyone know why the registerNewUser function was being called here?

crookedneighbor commented on test/api.mocha.coffee:
```diff
@@ -606,30 +607,30 @@ describe "API", ->
-      expect(user.purchased.plan.dateTerminated).to.not.be.ok()
-      done()
\ No newline at end of file
+      it "Handles unsubscription", (done) ->
```
Just needed to be indented to be in the correct block (note: Block set to skip right now)

crookedneighbor committed:
Added comment about expect

crookedneighbor committed:
Used a numTasks variable and added comments

crookedneighbor committed:
Added todo for area where test is failing to behave properly

crookedneighbor commented on test/api.mocha.coffee:
```diff
@@ -350,6 +356,11 @@ describe "API", ->
 
               # Send them invitations
               (_party, cb) ->
+                 # @TODO - _party should be array of newly
```
Could use some help here.

[These lines](https://github.com/HabitRPG/habitrpg/blob/develop/test/api.mocha.coffee#L340-L349) should register three new users, and when they're all done, the `_party` variable should be an array of user objects, the newly registered ones. Instead, it's an array of 3 user objects, all the the same user, the one used to make the api calls.

crookedneighbor replied:
Obviously, the invite calls don't work, because it's inviting the user who started the party to join the party (3 times)!

Alys replied:
Maybe it's meant to do that as a quick-and-dirty test to ensure that multiple user objects can be added to the party? From the point of view of actually adding users, it doesn't really matter if they're all the same user? Not ideal, but it might at least prove that arrays work. :)

crookedneighbor replied:
I don't think so, this is part of the preparation for testing quests. What is supposed to happen is that three users are invited to the party, they all accept, then they're invited to the quest and 2 accept and 1 doesn't.

I suspect the problem is with the registerNewUser function, somehow it's giving back the original user instead of a new one.

crookedneighbor replied:
After further investigation, turns out I was wrong, it's not the same user object. It's just that all the users have the same `_id`. 

(╯°□°）╯︵ ┻━┻

Alys replied:
Do the tests put the data into the production database?  I'm guessing not since the database is supposed to have protection against duplicate IDs, but I'm asking just in case. :)

crookedneighbor replied:
Tests get run against habitrpg_test, which is [recreated when the tests get started](https://github.com/HabitRPG/habitrpg/blob/develop/test/run_tests.sh#L12). 

I've never been able to actually explore the data saved to the test db though. Users never actually appear in the collection.

crookedneighbor replied:
Wait... it does create the user in the production db... WHY.

Alys replied:
Undesirable. That darned database is already huge. We have more than 30 users now. 

Thanks for mentioning that! It might explain some of the oddness I see in there, which I haven't finished investigating yet. You might have saved me some work. :)

crookedneighbor replied:
[This line is why](https://github.com/HabitRPG/habitrpg/blob/a8cb0fd4e9785f257ca5edbc9c62dba7577d5ca2/test/api.mocha.coffee#L21)

crookedneighbor replied:
@Alys Glad my hours of frustration resulted in something. :)

crookedneighbor replied:
This explains why A) my local install suddenly had 20+ users with admin capabilities and B) why the dread dragon kept appearing in the tavern.

crookedneighbor committed:
Updated request call to use correct form

crookedneighbor committed:
Wrote todo about what is wrong with the registerNewUser function

crookedneighbor committed:
Fixed typo in comment

crookedneighbor commented on test/api.mocha.coffee:
```diff
@@ -341,6 +347,14 @@ describe "API", ->
+                    # made. Instead of creating new users, it updates the original user
+                    # with a new login name, email and password. That is why
+                    # all the users have the same uuid, because they're not actually
+                    # being created. The original is just being updated. 
```
Is this how third party apps are supposed to update login name, email and password? By using the `/register` path and supplying the uuid and token of a user?

NickTomlin committed:
Fix nconf path in api test

crookedneighbor committed:
Moved karma dependency

NickTomlin committed:
WIP - Fix karma config file order to successfully boostrap app

There were a few issues here:

1. Relying on globs meant that modules were loaded out of order. So
'habitrpgc' was undefined when we were trying to add services etc to it.
The short term solution is to list them in the same way we list them in manifest
json. The long term would be switching everything to browserify so as to
avoid ordering/listing nightmare by having a single bundle.
2. We were missing exeternal deps (like 'hello') so the tests were
breaking before they even started. Some of them still are
3. We were including services that don't seem to exist anymore  (e.g. `habitrpgstatic`). We still are in certain places

NickTomlin committed:
Fix broken group and member service specs

Both were trying to declare an angular mock module that did not exist.
The group service spec urls were also asserting a trailing `?` that
doesn't seem to be party of the actual request.

lejeunerenard committed:
Make karma tests pass.

This pends a lot of tests that are failing due to translations not being loaded in the environment, as well as some issues with the page reload in groupServicesSpec.

We removed the module declarations for habitrpg in each spec since that
has been moved to specHelper.

crookedneighbor committed:
Merge pull request #4817 from lejeunerenard/fix-tests

WIP

Commit:
Skipping groups test in api for now, revisit later

lejeunerenard committed:
Add mongodb as a service for Travis-CI builds

crookedneighbor committed:
Corrected indentation of purchase test

lejeunerenard committed:
Use PhantomJS and run karma in Travis-CI

crookedneighbor committed:
Merge pull request #4818 from lejeunerenard/fix-tests

Fixed MongoDB service and added Karma tests to Travis-CI

crookedneighbor commented:
### TLDNR version

The travis build now passes. Overall, still a WIP, but I'd be in favor of merging in now so people can start writing tests to accompany their new features. Thoughts?

### Details

Travis yaml file fixed (mongo was never told to start up, which is why it was failing so quickly)

Karma configuration file is fixed, `grunt karma` can now run

The groups and purchasing api tests are still broken, I've set them to skip for now. Will continue to work on them.

The karma tests that interact with the translations strings are broken, and have been set to skip until they're fixed. 

The groupServices were causing a page refresh. Set those to skip until they can be fixed.

Alys commented:
:heart: Definitely merge now!

lemoness commented:
<3 <3 Thank you!!

On Sun, Mar 8, 2015 at 6:58 PM, Blade Barringer notifications@github.com
wrote:

> Merged #4671 https://github.com/HabitRPG/habitrpg/pull/4671.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/4671#event-248265772.
